File: frontend/lib/tradeLevels.ts
import type { ApiRecord, ScanPayload } from "@/lib/api";


const MAX_STOP_DISTANCE_PCT = 35;
const MAX_TARGET_DISTANCE_PCT = 75;
const MAX_ENTRY_DEVIATION_PCT = 25;
const INVALID_TRADE_LEVEL_STATE = "INVALID LEVELS — REVIEW REQUIRED";
const ACTIONABLE_TRADE_STATES = new Set([
  "ENTER NOW",
  "CONFIRMED BREAKOUT — BUY",
  "CONFIRMED BREAKDOWN — SELL"
]);
const BRACKET_FIELDS = ["entry_price", "target_price", "stop_loss_price"] as const;
const DISPLAY_LEVEL_FIELDS = ["entry", "target", "stop_loss"] as const;

type TradeLevelValidation = {
  valid: boolean;
  reason: string | null;
  risk_pct: number | null;
  reward_pct: number | null;
  entry_deviation_pct: number | null;
};

type SanitizedBracket = {
  record: ApiRecord;
  invalid: boolean;
  blockedActionable: boolean;
};

function finiteNumber(value: unknown): number | null {
  if (typeof value === "boolean" || value === null || value === undefined || value === "") return null;
  const number = Number(value);
  return Number.isFinite(number) ? number : null;
}

function roundMetric(value: number): number {
  return Math.round(value * 10_000) / 10_000;
}

export function validateLongTradeLevels({
  entryPrice,
  targetPrice,
  stopLossPrice,
  referencePrice
}: {
  entryPrice: unknown;
  targetPrice: unknown;
  stopLossPrice: unknown;
  referencePrice?: unknown;
}): TradeLevelValidation {
  const entry = finiteNumber(entryPrice);
  const target = finiteNumber(targetPrice);
  const stop = finiteNumber(stopLossPrice);
  if (entry === null || target === null || stop === null) {
    return {
      valid: false,
      reason: "Entry, target, and stop must all be finite numeric prices.",
      risk_pct: null,
      reward_pct: null,
      entry_deviation_pct: null
    };
  }
  if (!(0 < stop && stop < entry && entry < target)) {
    return {
      valid: false,
      reason: "Trade levels must satisfy 0 < stop < entry < target.",
      risk_pct: null,
      reward_pct: null,
      entry_deviation_pct: null
    };
  }

  const riskPct = ((entry - stop) / entry) * 100;
  const rewardPct = ((target - entry) / entry) * 100;
  if (riskPct > MAX_STOP_DISTANCE_PCT) {
    return {
      valid: false,
      reason: `Stop distance exceeds the ${MAX_STOP_DISTANCE_PCT}% safety limit.`,
      risk_pct: roundMetric(riskPct),
      reward_pct: roundMetric(rewardPct),
      entry_deviation_pct: null
    };
  }
  if (rewardPct > MAX_TARGET_DISTANCE_PCT) {
    return {
      valid: false,
      reason: `Target distance exceeds the ${MAX_TARGET_DISTANCE_PCT}% plausibility limit.`,
      risk_pct: roundMetric(riskPct),
      reward_pct: roundMetric(rewardPct),
      entry_deviation_pct: null
    };
  }

  const reference = finiteNumber(referencePrice);
  const entryDeviationPct =
    reference !== null && reference > 0
      ? (Math.abs(entry - reference) / reference) * 100
      : null;
  if (entryDeviationPct !== null && entryDeviationPct > MAX_ENTRY_DEVIATION_PCT) {
    return {
      valid: false,
      reason: `Entry price is more than ${MAX_ENTRY_DEVIATION_PCT}% from the reference price.`,
      risk_pct: roundMetric(riskPct),
      reward_pct: roundMetric(rewardPct),
      entry_deviation_pct: roundMetric(entryDeviationPct)
    };
  }

  return {
    valid: true,
    reason: null,
    risk_pct: roundMetric(riskPct),
    reward_pct: roundMetric(rewardPct),
    entry_deviation_pct: entryDeviationPct === null ? null : roundMetric(entryDeviationPct)
  };
}

function isActionable(record: ApiRecord): boolean {
  const state = String(record.trade_state ?? "").trim().toUpperCase();
  return record.is_actionable_now === true || ACTIONABLE_TRADE_STATES.has(state);
}

function sanitizeBracket(
  source: ApiRecord,
  referencePrice: unknown,
  explanationField: "explanation" | "trade_state_explanation"
): SanitizedBracket {
  const record: ApiRecord = { ...source };
  if (!BRACKET_FIELDS.some((field) => field in record)) {
    return { record, invalid: false, blockedActionable: false };
  }

  const values = BRACKET_FIELDS.map((field) => record[field]);
  const wasActionable = isActionable(record);
  if (values.every((value) => value === null || value === undefined) && !wasActionable) {
    return { record, invalid: false, blockedActionable: false };
  }

  const validation = validateLongTradeLevels({
    entryPrice: record.entry_price,
    targetPrice: record.target_price,
    stopLossPrice: record.stop_loss_price,
    referencePrice
  });
  record.trade_level_validation = validation;
  if (validation.valid) {
    return { record, invalid: false, blockedActionable: false };
  }

  const reason = validation.reason ?? "Trade levels failed validation.";
  for (const field of [...BRACKET_FIELDS, ...DISPLAY_LEVEL_FIELDS]) {
    if (field in record) record[field] = null;
  }
  record.action_block_reason = reason;
  if ("is_actionable_now" in record || wasActionable) record.is_actionable_now = false;
  if (wasActionable) {
    record.trade_state = INVALID_TRADE_LEVEL_STATE;
    record.trade_state_tone = "negative";
    record[explanationField] = reason;
    if ("recommendation_label" in record) record.recommendation_label = INVALID_TRADE_LEVEL_STATE;
    if ("tone" in record) record.tone = "negative";
  }
  return { record, invalid: true, blockedActionable: wasActionable };
}

function primaryHorizonKey(row: ApiRecord): "day_trade" | "swing_trade" | null {
  const label = String(
    row.primary_horizon_label
      ?? row.primary_horizon
      ?? row.expected_holding_period
      ?? row.ranking_bucket
      ?? ""
  ).toLowerCase();
  if (label.includes("5-15") || label.includes("swing")) return "swing_trade";
  if (label.includes("1-2") || label.includes("day")) return "day_trade";
  return null;
}

export function applyScanTradeLevelPolicy(payload: ScanPayload): ScanPayload {
  const invalidRowsBefore = finiteNumber(payload.market_stats?.invalid_trade_level_rows) ?? 0;
  const blockedBefore = finiteNumber(payload.market_stats?.blocked_invalid_level_signals) ?? 0;
  let invalidRowCount = 0;
  let blockedActionableCount = 0;

  const shortRows = (Array.isArray(payload.short_term) ? payload.short_term : []).map((sourceRow) => {
    let row: ApiRecord = { ...sourceRow };
    const referencePrice = row.current_price ?? row.entry_price;
    const primaryHorizon = primaryHorizonKey(row);
    const invalidHorizons = new Set<string>();
    let rowHadInvalidLevels = false;

    for (const horizonKey of ["day_trade", "swing_trade"] as const) {
      const horizon = row[horizonKey];
      if (!horizon || typeof horizon !== "object" || Array.isArray(horizon)) continue;
      const result = sanitizeBracket(horizon as ApiRecord, referencePrice, "explanation");
      row[horizonKey] = result.record;
      if (result.invalid) {
        invalidHorizons.add(horizonKey);
        rowHadInvalidLevels = true;
      }
    }

    const parentResult = sanitizeBracket(row, row.current_price, "trade_state_explanation");
    row = parentResult.record;
    rowHadInvalidLevels = rowHadInvalidLevels || parentResult.invalid;
    let parentBlocked = parentResult.blockedActionable;

    if (primaryHorizon !== null && invalidHorizons.has(primaryHorizon) && !parentResult.invalid) {
      const horizon = row[primaryHorizon] as ApiRecord;
      const reason = String(
        horizon.action_block_reason ?? "The primary trade horizon failed level validation."
      );
      const wasActionable = isActionable(row);
      for (const field of [...BRACKET_FIELDS, ...DISPLAY_LEVEL_FIELDS]) {
        if (field in row) row[field] = null;
      }
      row.action_block_reason = reason;
      row.is_actionable_now = false;
      row.trade_level_validation = {
        valid: false,
        reason,
        risk_pct: null,
        reward_pct: null,
        entry_deviation_pct: null
      };
      if (wasActionable) {
        row.recommendation_label = INVALID_TRADE_LEVEL_STATE;
        row.trade_state = INVALID_TRADE_LEVEL_STATE;
        row.trade_state_tone = "negative";
        row.trade_state_explanation = reason;
        row.tone = "negative";
        parentBlocked = true;
      }
    }

    if (rowHadInvalidLevels) invalidRowCount += 1;
    if (parentBlocked) blockedActionableCount += 1;
    return row;
  });

  return {
    ...payload,
    short_term: shortRows,
    market_stats: {
      ...(payload.market_stats ?? {}),
      invalid_trade_level_rows: Math.max(invalidRowCount, invalidRowsBefore),
      blocked_invalid_level_signals: Math.max(blockedActionableCount, blockedBefore)
    }
  };
}

File: domain/signals/trade_levels.py
from __future__ import annotations

from copy import deepcopy
from dataclasses import asdict, dataclass
from math import isfinite
from typing import Any


MAX_STOP_DISTANCE_PCT = 35.0
MAX_TARGET_DISTANCE_PCT = 75.0
MAX_ENTRY_DEVIATION_PCT = 25.0
INVALID_TRADE_LEVEL_STATE = "INVALID LEVELS — REVIEW REQUIRED"

_ACTIONABLE_TRADE_STATES = {
    "ENTER NOW",
    "CONFIRMED BREAKOUT — BUY",
    "CONFIRMED BREAKDOWN — SELL",
}
_BRACKET_FIELDS = ("entry_price", "target_price", "stop_loss_price")
_DISPLAY_LEVEL_FIELDS = ("entry", "target", "stop_loss")


@dataclass(frozen=True)
class TradeLevelValidation:
    valid: bool
    reason: str | None
    risk_pct: float | None = None
    reward_pct: float | None = None
    entry_deviation_pct: float | None = None

    def to_dict(self) -> dict[str, Any]:
        return asdict(self)


def _finite_float(value: Any) -> float | None:
    if isinstance(value, bool):
        return None
    try:
        number = float(value)
    except (TypeError, ValueError):
        return None
    return number if isfinite(number) else None


def validate_long_trade_levels(
    *,
    entry_price: Any,
    target_price: Any,
    stop_loss_price: Any,
    reference_price: Any = None,
    max_stop_distance_pct: float = MAX_STOP_DISTANCE_PCT,
    max_target_distance_pct: float = MAX_TARGET_DISTANCE_PCT,
    max_entry_deviation_pct: float = MAX_ENTRY_DEVIATION_PCT,
) -> TradeLevelValidation:
    """Validate a long trade bracket before it can be treated as actionable."""

    entry = _finite_float(entry_price)
    target = _finite_float(target_price)
    stop = _finite_float(stop_loss_price)
    if entry is None or target is None or stop is None:
        return TradeLevelValidation(
            valid=False,
            reason="Entry, target, and stop must all be finite numeric prices.",
        )
    if not 0 < stop < entry < target:
        return TradeLevelValidation(
            valid=False,
            reason="Trade levels must satisfy 0 < stop < entry < target.",
        )

    risk_pct = ((entry - stop) / entry) * 100.0
    reward_pct = ((target - entry) / entry) * 100.0
    if risk_pct > max(float(max_stop_distance_pct), 0.0):
        return TradeLevelValidation(
            valid=False,
            reason=f"Stop distance exceeds the {max_stop_distance_pct:g}% safety limit.",
            risk_pct=round(risk_pct, 4),
            reward_pct=round(reward_pct, 4),
        )
    if reward_pct > max(float(max_target_distance_pct), 0.0):
        return TradeLevelValidation(
            valid=False,
            reason=f"Target distance exceeds the {max_target_distance_pct:g}% plausibility limit.",
            risk_pct=round(risk_pct, 4),
            reward_pct=round(reward_pct, 4),
        )

    reference = _finite_float(reference_price)
    entry_deviation_pct = None
    if reference is not None and reference > 0:
        entry_deviation_pct = (abs(entry - reference) / reference) * 100.0
        if entry_deviation_pct > max(float(max_entry_deviation_pct), 0.0):
            return TradeLevelValidation(
                valid=False,
                reason=f"Entry price is more than {max_entry_deviation_pct:g}% from the reference price.",
                risk_pct=round(risk_pct, 4),
                reward_pct=round(reward_pct, 4),
                entry_deviation_pct=round(entry_deviation_pct, 4),
            )

    return TradeLevelValidation(
        valid=True,
        reason=None,
        risk_pct=round(risk_pct, 4),
        reward_pct=round(reward_pct, 4),
        entry_deviation_pct=round(entry_deviation_pct, 4) if entry_deviation_pct is not None else None,
    )


def _is_actionable(record: dict[str, Any]) -> bool:
    state = str(record.get("trade_state") or "").strip().upper()
    return record.get("is_actionable_now") is True or state in _ACTIONABLE_TRADE_STATES


def _has_bracket_fields(record: dict[str, Any]) -> bool:
    return any(field in record for field in _BRACKET_FIELDS)


def _sanitize_invalid_bracket(
    record: dict[str, Any],
    *,
    reference_price: Any = None,
    explanation_field: str,
) -> tuple[dict[str, Any], bool, bool]:
    sanitized = dict(record)
    if not _has_bracket_fields(sanitized):
        return sanitized, False, False

    values = [sanitized.get(field) for field in _BRACKET_FIELDS]
    was_actionable = _is_actionable(sanitized)
    if all(value is None for value in values) and not was_actionable:
        return sanitized, False, False

    validation = validate_long_trade_levels(
        entry_price=sanitized.get("entry_price"),
        target_price=sanitized.get("target_price"),
        stop_loss_price=sanitized.get("stop_loss_price"),
        reference_price=reference_price,
    )
    sanitized["trade_level_validation"] = validation.to_dict()
    if validation.valid:
        return sanitized, False, False

    reason = validation.reason or "Trade levels failed validation."
    for field in (*_BRACKET_FIELDS, *_DISPLAY_LEVEL_FIELDS):
        if field in sanitized:
            sanitized[field] = None
    sanitized["action_block_reason"] = reason
    if "is_actionable_now" in sanitized or was_actionable:
        sanitized["is_actionable_now"] = False
    if was_actionable:
        sanitized["trade_state"] = INVALID_TRADE_LEVEL_STATE
        sanitized["trade_state_tone"] = "negative"
        sanitized[explanation_field] = reason
        if "recommendation_label" in sanitized:
            sanitized["recommendation_label"] = INVALID_TRADE_LEVEL_STATE
        if "tone" in sanitized:
            sanitized["tone"] = "negative"
    return sanitized, True, was_actionable


def _primary_horizon_key(row: dict[str, Any]) -> str | None:
    label = str(
        row.get("primary_horizon_label")
        or row.get("primary_horizon")
        or row.get("expected_holding_period")
        or row.get("ranking_bucket")
        or ""
    ).lower()
    if "5-15" in label or "swing" in label:
        return "swing_trade"
    if "1-2" in label or "day" in label:
        return "day_trade"
    return None


def apply_scan_trade_level_policy(payload: dict[str, Any]) -> dict[str, Any]:
    """Return a scan payload with malformed or implausible trade levels blocked."""

    protected = deepcopy(payload)
    short_rows: list[dict[str, Any]] = []
    invalid_row_count = 0
    blocked_actionable_count = 0

    for source_row in protected.get("short_term", []):
        if not isinstance(source_row, dict):
            continue
        row = dict(source_row)
        reference_price = row.get("current_price") or row.get("entry_price")
        primary_horizon = _primary_horizon_key(row)
        invalid_horizons: set[str] = set()
        row_had_invalid_levels = False

        for horizon_key in ("day_trade", "swing_trade"):
            horizon = row.get(horizon_key)
            if not isinstance(horizon, dict):
                continue
            sanitized_horizon, invalid, _ = _sanitize_invalid_bracket(
                horizon,
                reference_price=reference_price,
                explanation_field="explanation",
            )
            row[horizon_key] = sanitized_horizon
            if invalid:
                invalid_horizons.add(horizon_key)
                row_had_invalid_levels = True

        row, parent_invalid, parent_blocked = _sanitize_invalid_bracket(
            row,
            reference_price=row.get("current_price"),
            explanation_field="trade_state_explanation",
        )
        row_had_invalid_levels = row_had_invalid_levels or parent_invalid
        primary_invalid = primary_horizon in invalid_horizons
        if primary_invalid and not parent_invalid:
            reason = str(
                row.get(primary_horizon, {}).get("action_block_reason")
                or "The primary trade horizon failed level validation."
            )
            was_actionable = _is_actionable(row)
            for field in (*_BRACKET_FIELDS, *_DISPLAY_LEVEL_FIELDS):
                if field in row:
                    row[field] = None
            row["action_block_reason"] = reason
            row["is_actionable_now"] = False
            row["trade_level_validation"] = {
                "valid": False,
                "reason": reason,
                "risk_pct": None,
                "reward_pct": None,
                "entry_deviation_pct": None,
            }
            if was_actionable:
                row["recommendation_label"] = INVALID_TRADE_LEVEL_STATE
                row["trade_state"] = INVALID_TRADE_LEVEL_STATE
                row["trade_state_tone"] = "negative"
                row["trade_state_explanation"] = reason
                row["tone"] = "negative"
                parent_blocked = True

        if row_had_invalid_levels:
            invalid_row_count += 1
        if parent_blocked:
            blocked_actionable_count += 1
        short_rows.append(row)

    stats = dict(protected.get("market_stats") or {})
    previous_invalid = stats.get("invalid_trade_level_rows")
    previous_blocked = stats.get("blocked_invalid_level_signals")
    stats["invalid_trade_level_rows"] = max(
        invalid_row_count,
        int(previous_invalid) if isinstance(previous_invalid, (int, float)) else 0,
    )
    stats["blocked_invalid_level_signals"] = max(
        blocked_actionable_count,
        int(previous_blocked) if isinstance(previous_blocked, (int, float)) else 0,
    )
    protected["market_stats"] = stats
    protected["short_term"] = short_rows
    return protected


__all__ = [
    "INVALID_TRADE_LEVEL_STATE",
    "MAX_ENTRY_DEVIATION_PCT",
    "MAX_STOP_DISTANCE_PCT",
    "MAX_TARGET_DISTANCE_PCT",
    "TradeLevelValidation",
    "apply_scan_trade_level_policy",
    "validate_long_trade_levels",
]
